samza-rest/src/main/java/org/apache/samza/rest/proxy/job/YarnRestJobStatusProvider.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.rest.proxy.job;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.codec.binary.StringUtils;
+import org.apache.commons.httpclient.HttpClient;
+import org.apache.commons.httpclient.HttpStatus;
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.hadoop.yarn.api.records.YarnApplicationState;
+import org.apache.samza.SamzaException;
+import org.apache.samza.rest.model.ApplicationInfo;
+import org.apache.samza.rest.model.ApplicationInfoWrapper;
+import org.apache.samza.rest.model.Job;
+import org.apache.samza.rest.model.JobStatus;
+import org.apache.samza.rest.resources.JobsResourceConfig;
+import org.apache.samza.rest.resources.YarnJobResourceConfig;
+import org.codehaus.jackson.map.DeserializationConfig;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of the {@link JobStatusProvider} that retrieves
+ * the job status from the YARN REST api.
+ */
+public class YarnRestJobStatusProvider implements JobStatusProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(YarnRestJobStatusProvider.class);
+  private final String _apiEndpoint;
+  private final HttpClient _httpClient;
+  private final ObjectMapper _objectMapper;
+
+  public YarnRestJobStatusProvider(JobsResourceConfig config) {
+    YarnJobResourceConfig yarnConfig = new YarnJobResourceConfig(config);
+    this._objectMapper = new ObjectMapper();
+    this._objectMapper.configure(DeserializationConfig.Feature.UNWRAP_ROOT_VALUE, true);
+    this._httpClient = new HttpClient();
+    this._apiEndpoint = String.format("http://%s/ws/v1/cluster/apps?states=RUNNING,NEW,NEW_SAVING,SUBMITTED,ACCEPTED",
```

Comment:
This is a bit fragile to changes in the YarnApplicationState enum. It hasn't changed since 2013, but it's still possible on a major version bump.

Is there a performance hit for requesting all states?
If so, maybe stick with this approach but verify that the YarnApplicationState doesn't have any new values unknown to this class (because if it does, they may not be handled properly).